core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java

```diff
@@ -95,30 +107,84 @@ public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand
 
    @Override
    public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) throws Throwable {
+      Map<Object, Object> originalMap = command.getMap();
+      ConsistentHash ch = dm.getConsistentHash();
+      Address localAddress = rpcManager.getAddress();
       if (ctx.isOriginLocal()) {
-         Set<Address> primaryOwners = new HashSet<Address>(command.getAffectedKeys().size());
-         for (Object k : command.getAffectedKeys()) {
-            primaryOwners.add(cdl.getPrimaryOwner(k));
+         List<NotifyingFuture<Object>> futures = new ArrayList<NotifyingFuture<Object>>(
+               rpcManager.getMembers().size() - 1);
+         // TODO: if async we don't need to do futures...
+         RpcOptions options = rpcManager.getDefaultRpcOptions(isSynchronous(command));
+         for (Address member : rpcManager.getMembers()) {
```

Comment:
To be honest both this and the next one you can try it either way, but normally I would think the number of entries would vastly outnumber the number of members or segments.  We could enhance it to do the inverse if one is smaller for example.